### DISCUSSION TITLE
POCSAG Midland SD225

### DISCUSSION THREAD

**Discussion by jgromes** [jgromes/RadioLib#601] [Q&A]
Following from #7 

@brooksbUWO what does your receiver chain look like? PDW takes input either from soundcard or from RS-232.

**Comment by brooksbUWO** [comment]
> Following from #7
> 
> @brooksbUWO what does your receiver chain look like? PDW takes input either from soundcard or from RS-232.

Soundcard with virtual audio cable.

**Reply by jgromes** [reply]
OK, so what is producing the sound? Some SDR software? And what demodulator?

**Reply by brooksbUWO** [reply]
Yes, some SDR software and specifically HDSDR with RTL2832U.

**Reply by jgromes** [reply]
And which mode are you using? NFM? SSB? Something else ...?

**Reply by brooksbUWO** [reply]
FM

Here's a screenshot incase there's any other settings you might want to know.
![2022-11-04 16_26_26-Window](https://user-images.githubusercontent.com/102610627/200077875-578944f0-478f-4aa8-8ec7-873eb2e5a9f0.png)

**Reply by jgromes** [reply]
Thanks, I downloaded HDSDR and played around with it for a little bit. I was able to decode RadioLib-transmitted POCSAG with PDW, as well as the audio recording you provided. Is the audio you sent from RadioLib, or from the original transmitter?

**Reply by brooksbUWO** [reply]
Audio was recorded with HDSDR and it was from transmitted using the Midland SD225.

**Reply by jgromes** [reply]
That's interesting, because I was able to decode it with PDW, so assuming the HDSDR settings used to record that and RadioLib POCSAG were the same, it should be compatible.

I would start by verifying transmission from RadioLib with PDW. If you can get the same decodes from RadioLib and Midland in PDW, then Midland isn't doing anything extra. It could also be an issue with freuency - not sure what is your transmitter at the RadioLib side, but if it's a cheap SX1278 (like mine), then it might require some manual frequency tuning, because the crystals on these are usually very inaccurate.

**Reply by brooksbUWO** [reply]
> That's interesting, because I was able to decode it with PDW, so assuming the HDSDR settings used to record that and RadioLib POCSAG were the same, it should be compatible.

The previous zip file of the Midland recording was not the entire RF signal, it was only the IF signal because even zipped it was too large since there is a 25MB file size limit here. The signal from the Midland and from SX1276 does not look the same. 

Just to make sure, I made another recording of the Midland transmitting tonight. Ever hour at the top of the hour, it sends the page twice and then a minute later it sends it again twice. I have a sign setup with the POCSAG receiver module attached and it received the signal and updated itself. I was not able to get PDW to decode it (signal level is 96.2% in PDW).

I also made a recording for the SX1276 sending a page. PDW was able to decode it, but not display the message. As I mentioned previously, I am only able to use one of the three methods to send a page with RadioLib (`PagerClient::transmit(uint8_t* data, size_t len, uint32_t addr, uint8_t encoding)`). I don't think PDW can display the uint8 data format. I can make the other 2 methods transmit by commenting out the following in Pager.cpp `if(((data == NULL) && (len > 0)) || ((data != NULL) && (len == 0))) { return(RADIOLIB_ERR_INVALID_PAYLOAD); }`

Here's the signal from Midland:
![2022-11-04 19_25_57-Window](https://user-images.githubusercontent.com/102610627/200092521-90ae586d-49cd-4d0d-92d0-b940e2660b14.png)

Here's a [link](https://drive.google.com/file/d/1KVkKDtzUdzgVvTvuJsIUewgAk16GYncZ/view?usp=sharing) to the full RF recording from Midland.

Here's the signal from SX1276:
![2022-11-04 19_41_24-Window](https://user-images.githubusercontent.com/102610627/200093264-1710b717-5bbe-4c05-990f-4624b3711865.png)

Here's a [link](https://drive.google.com/file/d/1w_b3HeypIhgjfCpnxjaW1uyXdK9vF0Tm/view?usp=sharing) to the full RF recording from SX1276.

**Reply by jgromes** [reply]
> As I mentioned previously, I am only able to use one of the three methods to send a page with RadioLib

Yes, and that is intentional. As I said previously, your data cointains NULLs, it cannot be trated as any sort of string. That's why the uint8_t method exists.

> I don't think PDW can display the uint8 data format.

It is not a special "format", every other method uses this method as well. I tried transmitting your binary test message and PDW receives the message, but displays nothing, so the next question is - are you sure it's really transmitting binary data, and not the literal string `~00~00~00~00` etc.? Because I was very easily able to achieve the same output as you by `transmit("~00~00~00~00")`

![Screenshot_35](https://user-images.githubusercontent.com/3407192/200112257-de661d10-7f1b-4336-b91c-123d75d0a55b.png)

Regarding different signals from RadioLib and Midland, it looks like frequency deviation from Midland is +-2250 Hz, whereas RadioLib has +-4500 Hz. I think the latter should be correct, since the standard specifies +-4.5 kHz deviation. However, out of curiousity I used 2250 Hz shifts in RadioLib and that also decodes in PDW.

**Reply by brooksbUWO** [reply]
> are you sure it's really transmitting binary data, and not the literal string `~00~00~00~00` etc.?

I am transmitting using the only one of your the three methods that will work which is `PagerClient::transmit(uint8_t* data, size_t len, uint32_t addr, uint8_t encoding)`. 

This is an example of the data being encoded using the Pager module:
`uint8_t command[] = {0x00, 0x00, 0x00, 0x00, 0x00, 0x01, 0x46, 0x46, 0x30, 0x30, 0x02, 0x41, 0x30, 0x43, 0x35, 0x30, 0x37, 0x46, 0x30, 0x30, 0x30, 0x30, 0x32, 0x33, 0x35, 0x39, 0xff, 0xff, 0xff, 0x33, 0xfe, 0x46, 0xfd, 0x42, 0xfa, 0x44, 0x03, 0x30, 0x61, 0x31, 0x66, 0x04};
`

**Reply by jgromes** [reply]
OK, but what is the Midland radio transmitting? Is it binary data or the string "~00~00~00~00"?

**Reply by brooksbUWO** [reply]
The Midland radio is transmitting the POCSAG encoded command. The data is captured using a RS232 monitoring cable by monitoring the input to the Midland's RS232 pin 1.

![2022-11-07 17_31_58-Midland sd-225 RS232 connector](https://user-images.githubusercontent.com/102610627/200437705-28f98556-355b-45e4-ab28-19fecbc551d9.png)


![PXL_20221107_231825699](https://user-images.githubusercontent.com/102610627/200436018-ce48aed6-7667-45b6-90d1-83d3a9c87edb.jpg)

Here's the captured data: 512 baud, 8 data bits, parity none, stop bits 1, none
`55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 3E 09 DD 78 91 30 F7 45 30 F7 45 30 F7 45 30 F7 45 30 F7 45 30 F7 45 30 F7 45 30 67 1C BF 50 18 56 9C 18 EE 99 18 80 9F 18 80 9B 18 B6 E6 D8 E8 CC 1A 0A 7D 09 DD 84 30 BC 37 0C D0 F2 7E 46 DB 00 C1 AF E4 98 F7 91 30 F7 45 30 F7 45 30 F7 45 30 F7 45 30 F7 45 30 F7 45 30 F7 45 30 F7 45 30 F7 45 30 F7 45 30 FF `

I used Bray's Terminal for the serial data capture.
![Bray's Terminal POCSAG capture from Midland Radio](https://user-images.githubusercontent.com/102610627/200439372-9601922c-4839-420f-960e-2127dfa4ab3c.png)

**Reply by jgromes** [reply]
That looks odd. The preamble should be 0xAA, not 0x55, so there might be some issues with endianness/MSB order. In addition, after preamble, there should be the frame sync word, 0x7CD215D8, but it's nowhere to be found.

**Reply by brooksbUWO** [reply]
I agree that it looks odd. The preamble should be 0xAA, but it is bit reversed and is 0x55. I don't have a data cable for the POCSAG receiver module, but there was a setting listing "Data Polarity". I don't know how the receiver's are actually programmed, but I'm starting to suspect they are inverted.

![BRG-1720 Programming Data Polarity](https://user-images.githubusercontent.com/102610627/200671840-0a633d2d-1e26-4f47-ba90-4c999e0fc105.png)

I'm not sure what the "Data Polarity" actually is suppose to mean. There is another [project](https://github.com/F5OEO/rpitx/blob/cce1fe6acf90d4d34ce304aed48fe80ec4ff51e7/src/pocsag/pocsag.cpp#L555) that has an option called "inverts the modulation polarity". 

Maybe this is the reason RadioLib pages are not being received by the [POCSAG data receiver](http://www.wirelessdevices.com.tw/Data%20sheet%20for%20Y1707.pdf)? I'm going to attempt building a data cable to see if I can read the current settings.

**Reply by jgromes** [reply]
I think you're trying to do too many things at the same time. I would start by examination of the recorded data - what has to be done to turn them into POCSAG messages? Flipping bits? Some MSB/LSB reordering?

Once you have a message that conforms to the POCSAG specification (0xAA preamble, 0x7CD215D8 frame sync etc.), you will know exactly what is the transmitter doing. You can then look inside the frame and determine whether the data is actually binary or not.

**Reply by hheinreich** [reply]
I already know the format of the data. The data command is exactly the same as if I physically connect to the sign's serial port and send a command using Bray's Terminal or ESP32 or if the wireless POCSAG is connected to the signs serial port as it is for normal operation. 

The command is what is being encoded into POCSAG and sent. Here's the command protocol.

https://www.brgprecision.com/pdffiles/Protocol12.pdf

My thought on temporarily reprogramming the POCSAG receiver module was to see if it could receive non-inverted POCSAG messages using RadioLib. The programming software is so old that I can't even get it to run on a virtual machine with Windows 2000. 

So as you suggested, I am going to look at the data the radio is transmitting to see if it's just inverted POCSAG.

**Comment by sp2ong** [comment]
POCSAG is wide used by ham radio in Europe it maybe you can find some information on 

https://hampager.de/dokuwiki/doku.php
